keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,461 @@
+
+In this KEP, we are proposing a way to handle node shutdown cases that are not detected by Kubelet's Node Shutdown Manager. The pods will be forcefully deleted in this case, trigger the deletion of the VolumeAttachments, and pods will move to a new running node so that application can continue to function.
+
+Similarly, this approach can also be applied to the case when a node is in a non-recoverable state such as the hardware failure or broken OS.
+
+### Use Cases
+
+* User wants a stateful workload to failover to another running node in the following cases with minimum intervention when it's safe to do so.
+  * The node is shutdown
+  * The node is in other non-recoverable cases:
+    * hardware failure
+    * broken OS
+
+### Goals
+
+* Increase the availability of stateful workloads when a node is shutdown, or in other non-recoverable cases such as the hardware failure or broken OS.
+* Automate self-healing for stateful workloads.
+
+### Non-Goals
+
+* Node/control plane partitioning other than a node shutdown is not covered by the proposal, but will be addressed in the future and built on top of this design.
+  * We do not have a way to determine whether it is safe to detach in the case of node partitioning. In the Alternatives section, we discussed the node fencing approach and explained why that is not selected for this KEP but will be considered after this KEP is implemented.
+* Implement in-cluster logic to handle node/control plane partitioning.
+
+## Proposal
+
+In this section, we are describing the proposed changes to enable workloads to failover to a running node when the original node is shutdown but not detected by Kubelet's Node Shutdown Manager, or when the node is in other non-recoverable cases such as the hardware failure or broken OS.
+
+Existing logic:
+1. When a node is not reachable from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.
+
+1. After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.
+
+Proposed logic change:
+1. [Proposed change] This proposal requires a user to apply a `out-of-service` taint on a node when the user has confirmed that this node is shutdown or in a non-recoverable state due to the hardware failure or broken OS. Note that user should only add this taint if the node is not coming back at least for some time. If the node is in the middle of restarting, this taint should not be used.
+
+1. [Proposed change] In the Pod GC Controller, part of the kube-controller-manager, add a new function called gcTerminating. This function would need to go through all the Pods in terminating state, verify that the Node is NotReady. If so, do the following:
```

Comment:
Yes, NoExecuteTaintManager tries to delete Pods already in existing code as there's already a NoExecute taint on the Pod when node becomes Not Ready. The TaintManager will try to delete the Pod if there is "at least" one NoExecute taint.  So adding another out-of-service NoExecute taint will not change this behavior. We do not need to change the code in TaintManager.

The Pods will be stuck in terminating status as described in the KEP. They will be forcefully deleted by the Pod GC controller as proposed in the KEP.

Reply:
I think it may be cleaner to have all logic handling NoExecute taint to be in the same controller, wdyt?

Reply:
Pod GC Controller is the only controller that forcefully deletes pods (grace period = 0). This is probably on purpose because it does garbage collection.  I'm not sure if we should let NoExecuteTaintManager also forcefully deletes pods.

Reply:
[Taints and Tolerations](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#concepts) says:

> Normally, if a taint with effect NoExecute is added to a node, then any pods that do not tolerate the taint will be evicted **immediately**, and pods that do tolerate the taint will never be evicted.

So, Taint Manager also deletes pods forcefully if it meant no grace period.

**But**, I'm ok that Pod GC Controller will do it because stalled terminating pods are just garbage.

Reply:
In Pod GC Controller, it uses this NewDeleteOptions{0}. That means grace period is set to 0.
https://github.com/kubernetes/kubernetes/blob/v1.24.0-alpha.1/pkg/controller/podgc/gc_controller.go#L79

In TaintManager, it uses DeleteOptions{}. That means default grace period.
https://github.com/kubernetes/kubernetes/blob/v1.24.0-alpha.1/pkg/controller/nodelifecycle/scheduler/taint_manager.go#L112

I searched for NewDeleteOptions{0} and only found it is used in Pod GC Controller. I also don't see DeleteOptions with GracePeriodSeconds set to 0 directly.
https://github.com/kubernetes/kubernetes/blob/v1.24.0-alpha.1/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/helpers.go#L228

Reply:
I see. Thank you.